### PR TITLE
GH-2307 Adds GraphC

### LINKED ISSUE #2307: Canvas Graph widget for large relational data sets

### PR DESCRIPTION
**Pull request by jbrundage**
Fixes #2307 

@GordonSmith please review

### CODE REVIEW

**Comment by jbrundage** [comment]
@GordonSmith This is revised and ready to be reviewed

File: src/graph/VertexC.js [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+        if (this.min_y > _y) this.min_y = _y;
+        if (this.max_y < _y) this.max_y = _y;
+    };
+
+    VertexC.prototype.getBBox = function () {
+        return this.size();
+    };
+
+    //  Render  ---
+    VertexC.prototype.enter = function (domNode, element) {
```
Comment (GordonSmith) [inline src/graph/VertexC.js]: Not used?

File: src/graph/VertexC.js [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+        SVGWidget.prototype.enter.apply(this, arguments);
+        this._icon
+            .target(domNode)
+            .render();
+        this._textBox
+            .target(domNode)
+            .render();
+    };
+
+    VertexC.prototype.update = function (domNode, element) {
```
Comment (GordonSmith) [inline src/graph/VertexC.js]: Not used?

File: src/graph/VertexC.js [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+        annotations.exit()
+            .each(function (d, idx) {
+                var element = d3.select(this);
+                delete context._annotationWidgets[idx];
+                element.remove();
+            });
+    };
+
+    //  Methods  ---
+    VertexC.prototype.intersection = function (pointA, pointB) {
```
Comment (GordonSmith) [inline src/graph/VertexC.js]: Not used?

File: src/graph/VertexC.js [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+    VertexC.prototype.publish("iconTooltip", "", "string", "iconTooltip", null, {tags: ["Private"]});
+
+    VertexC.prototype.publish("annotationDiameter", 14, "number", "Annotation Diameter", null, {tags: ["Private"]});
+    VertexC.prototype.publish("annotationSpacing", 3, "number", "Annotation Spacing", null, {tags: ["Private"]});
+    VertexC.prototype.publish("annotationIcons", [], "array", "Annotations", null, {tags: ["Private"]});
+
+    VertexC.prototype.drawOriginalLayout = function (ctx, canvas) {
+        //TODO
+    };
+    VertexC.prototype.drawCenteredLayout = function (ctx, canvas) {
```
Comment (GordonSmith) [inline src/graph/VertexC.js]: Not used?

File: src/graph/VertexC.js [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+    VertexC.prototype.publish("iconAnchor", "start", "set", "Icon Anchor Position", ["", "start", "middle", "end"], {tags: ["Basic"]});
+
+    VertexC.prototype.publish("tooltip", "", "string", "Tooltip", null, {tags: ["Private"]});
+    VertexC.prototype.publish("iconTooltip", "", "string", "iconTooltip", null, {tags: ["Private"]});
+
+    VertexC.prototype.publish("annotationDiameter", 14, "number", "Annotation Diameter", null, {tags: ["Private"]});
+    VertexC.prototype.publish("annotationSpacing", 3, "number", "Annotation Spacing", null, {tags: ["Private"]});
+    VertexC.prototype.publish("annotationIcons", [], "array", "Annotations", null, {tags: ["Private"]});
+
+    VertexC.prototype.drawOriginalLayout = function (ctx, canvas) {
```
Comment (GordonSmith) [inline src/graph/VertexC.js]: Not used?

File: src/graph/VertexC.js [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+    VertexC.prototype.publishProxy("textbox_text_colorFill", "_textBox", "text_colorFill");
+
+    VertexC.prototype.publish("iconAnchor", "start", "set", "Icon Anchor Position", ["", "start", "middle", "end"], {tags: ["Basic"]});
+
+    VertexC.prototype.publish("tooltip", "", "string", "Tooltip", null, {tags: ["Private"]});
+    VertexC.prototype.publish("iconTooltip", "", "string", "iconTooltip", null, {tags: ["Private"]});
+
+    VertexC.prototype.publish("annotationDiameter", 14, "number", "Annotation Diameter", null, {tags: ["Private"]});
+    VertexC.prototype.publish("annotationSpacing", 3, "number", "Annotation Spacing", null, {tags: ["Private"]});
+    VertexC.prototype.publish("annotationIcons", [], "array", "Annotations", null, {tags: ["Private"]});
```
Comment (GordonSmith) [inline src/graph/VertexC.js]: Can you remove any PP not used?

File: src/graph/EdgeC.js [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+    EdgeC.prototype._class += " graph_EdgeC";
+
+    EdgeC.prototype.publish("arcDepth", 16, "number", "Arc Depth", null, { tags: ["Basic"] });
+    EdgeC.prototype.publish("showArc", true, "boolean", "Show/Hide Arc", null, { tags: ["Basic"] });
+    EdgeC.prototype.publish("tooltip", "", "string", "Tooltip", null, { tags: ["Private"] });
+
+    EdgeC.prototype.publish("sourceMarker", "circle", "set", "Source Marker", ["circle"], { optional: true });
+    EdgeC.prototype.publish("targetMarker", "arrow", "set", "Source Marker", ["arrow", "circle"], { optional: true });
+    EdgeC.prototype.publish("strokeDasharray", null, "string", "Stroke Dash Array", null, { optional: true });
+    EdgeC.prototype.publish("strokeColor", null, "html-color", "Stroke Color", null, { optional: true });
```
Comment (GordonSmith) [inline src/graph/EdgeC.js]: Can you remove any unused PP?

File: src/graph/GraphC.js [hunk 8]
```diff
@@ -0,0 +1,809 @@
...
+            var k_scale = this.zoom.scale();
+            var w = _canvas.width / k_scale;
+            var h = _canvas.height / k_scale;
+            //TODO: figure out why this only clears bottom right occasionally context.ctx.clearRect(-x * 10, -y * 10, w * 10, h * 10);
+            // context.ctx.clearRect(-x * 10, -y * 10, w * 10, h * 10);
+            context.ctx.clearRect(-w * 10, -h * 10, w * 10 * 2, h * 10 * 2);
+        }
+    };
+    //  Properties  ---
+    GraphC.prototype.getOffsetPos = function () {
```
Comment (GordonSmith) [inline src/graph/GraphC.js]: Used?

File: src/graph/GraphC.js [hunk 9]
```diff
@@ -0,0 +1,809 @@
...
+                switch (context._zoomMode) {
+                    case "selection":
+                        break;
+                    default:
+                        context.applyZoom();
+                        break;
+                }
+                context.draw();
+            });
+        this.brush = d3.svg.brush()
```
Comment (GordonSmith) [inline src/graph/GraphC.js]: Is brushing working?  If not remove.

File: src/graph/GraphC.js [hunk 10]
```diff
@@ -0,0 +1,809 @@
...
+            }
+        }
+
+        return {
+            vertices: vertices,
+            edges: edges
+        };
+    };
+
+    GraphC.prototype.highlightVerticies = function (vertexMap) {
```
Comment (GordonSmith) [inline src/graph/GraphC.js]: Used?

File: src/graph/GraphC.js [hunk 11]
```diff
@@ -0,0 +1,809 @@
...
+            .style("opacity", function (d) {
+                if (!vertexMap || vertexMap[d.id()]) {
+                    return 1;
+                }
+                return context.highlight.opacity;
+            });
+        return this;
+    };
+
+    GraphC.prototype.highlightEdges = function (edgeMap) {
```
Comment (GordonSmith) [inline src/graph/GraphC.js]: Used?

File: src/graph/GraphC.js [hunk 12]
```diff
@@ -0,0 +1,809 @@
...
+            .style("opacity", function (o) {
+                if (!edgeMap || edgeMap[o.id()]) {
+                    return 1;
+                }
+                return context.highlight.opacity;
+            });
+        return this;
+    };
+
+    GraphC.prototype.highlightVertex = function (element, d) {
```
Comment (GordonSmith) [inline src/graph/GraphC.js]: Used?

File: src/graph/GraphC.js [hunk 13]
```diff
@@ -0,0 +1,809 @@
...
+                this.highlightVerticies(highlight.vertices);
+                this.highlightEdges(highlight.edges);
+            } else {
+                this.highlightVerticies(null);
+                this.highlightEdges(null);
+            }
+        }
+    };
+
+    GraphC.prototype.highlightEdge = function (element, d) {
```
Comment (GordonSmith) [inline src/graph/GraphC.js]: Used?

**Pull request merged**